app/models/miq_event.rb

```diff
@@ -34,13 +30,42 @@ def self.description
     _("Policy Events")
   end
 
+  def self.group_name(group)
+    return if group.nil?
+
+    group_names_and_levels.dig(:group_names, group.to_sym) || DEFAULT_GROUP_NAME_STR
+  end
+
+  def self.group_and_level(event_type)
+    by_literal = partition_group_and_level_by_event_type
+    by_literal[event_type.to_s] || [DEFAULT_GROUP_NAME, DEFAULT_GROUP_LEVEL]
+  end
+
+  private_class_method def self.partition_group_and_level_by_event_type
+    return @literal_group_and_level_by_event_type if @literal_group_and_level_by_event_type
+
+    @literal_group_and_level_by_event_type = {}
+
+    # TODO: check if there's a more performant way to look up the miq_sets and miq_set_membership
+    @literal_group_and_level_by_event_type = MiqEventDefinition.all.each_with_object({}) do |ed, h|
+      group_name = ed.event_group_name || DEFAULT_GROUP_NAME
+      h[ed.name] = [group_name.to_sym, DEFAULT_GROUP_LEVEL]
+    end
+  end
+
+  def self.clear_event_groups_cache
+    @group_names_and_levels, @literal_group_and_level_by_event_type = nil
+  end
+
   def self.group_names_and_levels
-    hash = default_group_names_and_levels
-    hash[:group_names].merge!(MiqEventDefinitionSet.all.pluck(:name, :description).to_h.symbolize_keys)
-    group_levels.each do |level|
-      hash[:group_levels][level] ||= level.to_s.capitalize
+    @group_names_and_levels ||= begin
+      hash = default_group_names_and_levels
+      hash[:group_names].merge!(MiqEventDefinitionSet.all.pluck(:name, :description).to_h.symbolize_keys)
```

Comment:
If there are new event definitions, do we bust this cache?

Reply:
No.  You'd need to seed to get them. We can bust it but I had no idea when you'd do that.

Reply:
If it's seeded then there's no need to bust it, because reseed only happens after an upgrade / restart.